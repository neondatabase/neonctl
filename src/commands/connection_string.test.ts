import { describe, expect } from 'vitest';
import { test } from '../test_utils/fixtures';

describe('connection_string', () => {
  test('connection_string', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
    ]);
  });

  test('connection_string branch id', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'br-sunny-branch-123456',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
    ]);
  });

  test('connection_string branch id 8 digits', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'br-cloudy-branch-12345678',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
    ]);
  });

  test('connection_string pooled', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--pooled',
    ]);
  });

  test('connection_string prisma', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--prisma',
    ]);
  });

  test('connection_string prisma pooled', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--prisma',
      '--pooled',
    ]);
  });

  test('connection_string prisma pooled extended', async ({
    testCliCommand,
  }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--prisma',
      '--pooled',
      '--extended',
    ]);
  });

  test('connection_string without any args should pass', async ({
    testCliCommand,
  }) => {
    await testCliCommand(['connection-string'], {
      mockDir: 'single_project',
    });
  });

  test('connection_string with psql', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--psql',
    ]);
  });

  test('connection_string with psql and psql args', async ({
    testCliCommand,
  }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--psql',
      '--',
      '-c',
      'SELECT 1',
    ]);
  });

  test('connection_string without ssl', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--ssl',
      'omit',
    ]);
  });

  test('connection_string with ssl verify full', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
      '--ssl',
      'verify-full',
    ]);
  });

  test('connection_string with lsn', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch@0/123456',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
    ]);
  });

  test('connection_string with timestamp', async ({ testCliCommand }) => {
    await testCliCommand([
      'connection-string',
      'test_branch@2021-01-01T00:00:00Z',
      '--project-id',
      'test',
      '--database-name',
      'test_db',
      '--role-name',
      'test_role',
    ]);
  });

  test('connection_string fails for non-existing database', async ({
    testCliCommand,
  }) => {
    await testCliCommand(
      [
        'connection-string',
        'test_branch',
        '--project-id',
        'test',
        '--database-name',
        'non_existing_db',
        '--role-name',
        'test_role',
      ],
      {
        code: 1,
        stderr: expect.stringMatching(/Database not found/),
      },
    );
  });
});
