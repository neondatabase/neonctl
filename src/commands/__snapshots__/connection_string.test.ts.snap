// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connection_string > connection_string 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db?sslmode=require
"
`;

exports[`connection_string > connection_string branch id 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db?sslmode=require
"
`;

exports[`connection_string > connection_string branch id 8 digits 1`] = `
"postgresql://test_role:test_pwd@ep-lame-king-930914.us-east-2.aws.neon.tech/test_db?sslmode=require
"
`;

exports[`connection_string > connection_string fails for non-existing database 1`] = `""`;

exports[`connection_string > connection_string pooled 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914-pooler.us-east-2.aws.neon.tech/test_db?sslmode=require
"
`;

exports[`connection_string > connection_string prisma 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db?connect_timeout=30&sslmode=require
"
`;

exports[`connection_string > connection_string prisma pooled 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914-pooler.us-east-2.aws.neon.tech/test_db?connect_timeout=30&pool_timeout=30&pgbouncer=true&sslmode=require
"
`;

exports[`connection_string > connection_string prisma pooled extended 1`] = `
"connection_string: postgresql://test_role:test_pwd@ep-cool-king-930914-pooler.us-east-2.aws.neon.tech/test_db?connect_timeout=30&pool_timeout=30&pgbouncer=true&sslmode=require
host: ep-cool-king-930914-pooler.us-east-2.aws.neon.tech
role: test_role
password: test_pwd
database: test_db
options: connect_timeout=30&pool_timeout=30&pgbouncer=true&sslmode=require
"
`;

exports[`connection_string > connection_string with lsn 1`] = `
"postgresql://test_role:test_pwd@br-sunny-branch-123456.us-east-2.aws.neon.tech/test_db?sslmode=require&options=neon_lsn%3A0%2F123456
"
`;

exports[`connection_string > connection_string with psql 1`] = `"{"psql-cli-args":["postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db?sslmode=require"]}"`;

exports[`connection_string > connection_string with psql and psql args 1`] = `"{"psql-cli-args":["postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db?sslmode=require","-c","SELECT 1"]}"`;

exports[`connection_string > connection_string with ssl verify full 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db?sslmode=verify-full
"
`;

exports[`connection_string > connection_string with timestamp 1`] = `
"postgresql://test_role:test_pwd@br-sunny-branch-123456.us-east-2.aws.neon.tech/test_db?sslmode=require&options=neon_timestamp%3A2021-01-01T00%3A00%3A00Z
"
`;

exports[`connection_string > connection_string without any args should pass 1`] = `
"postgresql://test_role:password@undefined/db1?sslmode=require
"
`;

exports[`connection_string > connection_string without ssl 1`] = `
"postgresql://test_role:test_pwd@ep-cool-king-930914.us-east-2.aws.neon.tech/test_db
"
`;
